File: routes/product.js
const express = require("express");
const router = express.Router();
const Product = require("../models/product");

// Get all products
router.get("/", async (req, res) => {
  try {
    const products = await Product.find();
    res.json(products);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});


// Delete ALL products
router.delete("/", async (req, res) => {
  console.log("DELETE ALL request received");
  try {
    const result = await Product.deleteMany({});
    console.log(`Deleted ${result.deletedCount} products`);
    res.json({ message: `Deleted ${result.deletedCount} products` });
  } catch (err) {
    console.error("Error deleting all products:", err);
    res.status(500).json({ message: "Server Error: " + err.message });
  }
});

// Add a new product
router.post("/", async (req, res) => {
  const product = new Product({
    name: req.body.name,
    price: req.body.price,
    description: req.body.description,
    category: req.body.category,
    imageUrl: req.body.imageUrl,
    stock: req.body.stock,
  });

  try {
    const newProduct = await product.save();
    res.status(201).json(newProduct);
  } catch (err) {
    res.status(400).json({ message: err.message });
  }
});

// Update a product
router.put("/:id", async (req, res) => {
  console.log("PUT request received for ID:", req.params.id);
  console.log("Request Body:", req.body);
  try {
    const updatedProduct = await Product.findByIdAndUpdate(
      req.params.id,
      req.body,
      { new: true, runValidators: true }
    );
    if (!updatedProduct) {
      console.log("Product not found with ID:", req.params.id);
      return res.status(404).json({ message: "Product not found" });
    }
    console.log("Product updated successfully:", updatedProduct);
    res.json(updatedProduct);
  } catch (err) {
    console.error("Error updating product:", err);
    res.status(500).json({ message: "Server Error: " + err.message });
  }
});

// Delete a product
router.delete("/:id", async (req, res) => {
  console.log("DELETE request received for ID:", req.params.id);
  try {
    const deletedProduct = await Product.findByIdAndDelete(req.params.id);
    if (!deletedProduct) {
      console.log("Product not found for deletion:", req.params.id);
      return res.status(404).json({ message: "Product not found" });
    }
    console.log("Product deleted:", req.params.id);
    res.json({ message: "Product deleted" });
  } catch (err) {
    console.error("Error deleting product:", err);
    res.status(500).json({ message: "Server Error: " + err.message });
  }
});

module.exports = router;

File: Downloads/botclub-main (2)/botclub-main/flask_backend/routes/product.py
from flask import Blueprint, request, jsonify
from models.product import Product
from mongoengine import DoesNotExist, ValidationError
import json

product_bp = Blueprint('products', __name__)

def product_to_dict(product):
    return {
        "_id": str(product.id),
        "name": product.name,
        "price": product.price,
        "description": product.description,
        "category": product.category,
        "imageUrl": product.imageUrl,
        "stock": product.stock
    }

# Get all products
@product_bp.route('/', methods=['GET'])
def get_products():
    try:
        products = Product.objects()
        return jsonify([product_to_dict(p) for p in products]), 200
    except Exception as e:
        return jsonify({"message": str(e)}), 500

# Delete ALL products
@product_bp.route('/', methods=['DELETE'])
def delete_all_products():
    print("DELETE ALL request received")
    try:
        count = Product.objects.count()
        Product.objects.delete()
        print(f"Deleted {count} products")
        return jsonify({"message": f"Deleted {count} products"}), 200
    except Exception as e:
        print(f"Error deleting all products: {e}")
        return jsonify({"message": f"Server Error: {str(e)}"}), 500

# Add a new product
@product_bp.route('/', methods=['POST'])
def create_product():
    try:
        data = request.get_json()
        product = Product(
            name=data['name'],
            price=data['price'],
            description=data.get('description'),
            category=data.get('category'),
            imageUrl=data.get('imageUrl'),
            stock=data.get('stock', 0)
        )
        product.save()
        return jsonify(product_to_dict(product)), 201
    except Exception as e:
        return jsonify({"message": str(e)}), 400

# Update a product
@product_bp.route('/<id>', methods=['PUT'])
def update_product(id):
    print("PUT request received for ID:", id)
    print("Request Body:", request.get_json())
    try:
        data = request.get_json()
        product = Product.objects.get(id=id)
        
        # update fields manually or iterate
        for key, value in data.items():
            if hasattr(product, key):
                setattr(product, key, value)
        
        product.save()
        print("Product updated successfully:", product.to_json())
        return jsonify(product_to_dict(product)), 200
    except DoesNotExist:
        print("Product not found with ID:", id)
        return jsonify({"message": "Product not found"}), 404
    except Exception as e:
        print("Error updating product:", e)
        return jsonify({"message": f"Server Error: {str(e)}"}), 500

# Delete a product
@product_bp.route('/<id>', methods=['DELETE'])
def delete_product(id):
    print("DELETE request received for ID:", id)
    try:
        product = Product.objects.get(id=id)
        product.delete()
        print("Product deleted:", id)
        return jsonify({"message": "Product deleted"}), 200
    except DoesNotExist:
        print("Product not found for deletion:", id)
        return jsonify({"message": "Product not found"}), 404
    except Exception as e:
        print("Error deleting product:", e)
        return jsonify({"message": f"Server Error: {str(e)}"}), 500
